shared/src/token_info.rs

```diff
@@ -0,0 +1,97 @@
+use async_trait::async_trait;
+use contracts::ERC20;
+use ethcontract::{batch::CallBatch, Http, Web3, H160};
+use futures::future::join_all;
+use std::collections::HashMap;
+use std::sync::Arc;
+use tokio::sync::Mutex;
+
+use mockall::*;
+
+const MAX_BATCH_SIZE: usize = 100;
+
+#[cfg_attr(test, derive(Eq, PartialEq))]
+#[derive(Copy, Clone, Debug)]
+pub struct TokenInfo {
+    pub decimals: Option<u8>,
+}
+
+pub struct TokenInfoFetcher {
+    pub web3: Web3<Http>,
+}
+
+#[automock]
+#[async_trait]
+pub trait TokenInfoFetching: Send + Sync {
+    /// Retrieves all token information.
+    /// Default implementation calls get_token_info for each token and ignores errors.
+    async fn get_token_infos(&self, addresses: &[H160]) -> HashMap<H160, TokenInfo>;
+}
+
+#[async_trait]
+impl TokenInfoFetching for TokenInfoFetcher {
+    async fn get_token_infos(&self, addresses: &[H160]) -> HashMap<H160, TokenInfo> {
+        let web3 = Web3::new(self.web3.transport().clone());
+        let mut batch = CallBatch::new(self.web3.transport());
+        let futures = addresses
+            .iter()
+            .map(|address| {
+                let erc20 = ERC20::at(&web3, *address);
+                erc20.methods().decimals().batch_call(&mut batch)
+            })
+            .collect::<Vec<_>>();
+
+        batch.execute_all(MAX_BATCH_SIZE).await;
+
+        addresses
+            .iter()
+            .zip(join_all(futures).await.into_iter())
+            .map(|r| {
+                if (r.1.is_err()) {
```

Comment:
This is a bit risky. If there was a transient error (e.g. the node shutting down or a timeout), we will cache the token as not having any decimals until we restart. For the time being it might be better to not cache errors and create a TODO that we could inspect the error to decide if it's transient or not.

Reply:
You're right - changed it.